src/GraphQL.Tests/Types/SchemaTests.cs

```diff
@@ -160,6 +160,22 @@ private void DoesNotContainTypeNames(Schema schema, params string[] typeNames)
                 type.ShouldBe(null, $"Found {typeName} in type lookup.");
             });
         }
+
+        [Fact]
+        public void middleware_can_reference_SchemaTypes()
+        {
+            var schema = new Schema() { Query = new SomeQuery() };
+            schema.FieldMiddleware.Use((schema, next) =>
+            {
+                schema.AllTypes.Count.ShouldNotBe(0);
+                return async context =>
+                {
+                    var res = await next(context);
+                    return "One " + res.ToString();
+                };
+            });
+            schema.Initialize();
+        }
```

Comment:
If `schema` was only used within the actual field resolver, it could be pulled from the execution context.  So `schema` must only be necessary to build the middleware.

If I remember correctly, I think you added `schema` to the builder primarily so that it could be used to resolve a scoped middleware instance while building a scoped schema with the `Use<MyMiddleware>` syntax.  If that's the only reason for it, we could potentially replace `schema` with `serviceProvider`.  With `Middleware` being defined on the schema, maybe there's another way to handle this scenario.

Reply:
> If I remember correctly, I think you added schema to the builder primarily so that it could be used to resolve a scoped middleware instance while building a scoped schema with the Use<MyMiddleware> syntax.

Yes, exactly. But now I saw that the schema can be useful on its own when **constructiong** delegate. 

> If that's the only reason for it, we could potentially replace schema with serviceProvider.

50/50.  On the one hand, I see cases where the caller can get a reference to schema through a closure without any problems. On the other hand, there are probably cases when it is not so easy. Now the schema for building a middleware is, in fact, an auxiliary parameter that is not required in 100% of cases.

Reply:
I guess the middleware can be pulled from the schema constructor for DI injection. That makes sense to me.

Reply:
Do you suggest to get rid of `FieldMiddlewareBuilder` ?

Reply:
Or let's take `IFieldMiddlewareBuilder` as schema ctor.

Reply:
As I understand it, now there is no special sense in its existence as a property of the schema after it is initialized. It is needed at creation - exactly as a parameter of the constructor. You don't even need to store it.

Reply:
But not. It needs to be stored until the schema is initialized, but then it is no longer needed.

Reply:
What I meant is that it is now easy to define middleware like this:

```cs
public MySchema : Schema
{
    public MySchema(IServiceProvider services, MyQuery query, MyMutation mutation, MyMiddleware middleware) : base(services)
    {
        Query = query;
        Mutation = mutation;
        Middleware.Use(middleware);
    }
}
```

Reply:
Ok. Nice to document.

Reply:
I already did so. See updated middleware docs

Reply:
OK